### PR TITLE
Isaacs/set envs fix

### PR DESCRIPTION
**Pull request by isaacs**
Implement the changes in https://github.com/npm/rfcs/pull/183, and set environment variables properly when they are not strings.

Also adds tests for the set-envs util.

### CODE REVIEW

**Review by MylesBorins — APPROVED** [review]
Rubber Stamp LGTM

Tested locally and it fixes the bug I was running into re: missing env var

File: lib/config/set-envs.js [hunk 1]
```diff
@@ -8,11 +8,14 @@
...
   return !/^[/@_]/.test(key) &&
-    (typeof val === 'string' || Array.isArray(val)) &&
-    `npm_config_${key.replace(/-/g, '_').toLowerCase()}`
+    (typeof envVal(val) === 'string') &&
+      `npm_config_${key.replace(/-/g, '_').toLowerCase()}`
 }
 
-const envVal = val => Array.isArray(val) ? val.join('\n\n') : val
+const envVal = val => Array.isArray(val) ? val.map(v => envVal(v)).join('\n\n')
+  : val === null || val === undefined || val === false ? ''
```
Comment (darcyclarke) [inline lib/config/set-envs.js:16]: o_O can we at least start nesting these? ex.
```
const envVal = val => Array.isArray(val) ? val.map(v => envVal(v)).join('\n\n')
  : val === null || val === undefined || val === false ? ''
    : typeof val === 'object' ? null
      : String(val)
```

**Review by darcyclarke — APPROVED** [review]

File: lib/config/set-envs.js (hunk 1 above)
Comment (isaacs) [reply lib/config/set-envs.js:16]: We've moved to lining up the `:` for easier scanning down a straight left-edge, so the CLI matches all the other repos.

**Comment by isaacs** [comment]
Landed and published on 7.0.0-beta.4.

**Pull request closed without merging**